integration-test-utils/src/main/java/com/palantir/gradle/util/TestDepVersions.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Utility class to keep versions of dependencies referenced in test files up to date with the versions declared in
+ * the project.  The class looks both at versions.lock and versions.props files to resolve dependencies.
+ *
+ * {@code
+ *    import static com.palantir.test.TestDepVersions.resolve
+ *    //...within a test case...
+ *      buildFile << """
+ *         buildscript {
+ *             repositories {
+ *                 maven { url 'https://artifactory.palantir.build/artifactory/release-jar' }
+ *             }
+ *             dependencies {
+ *                 classpath '${resolve('com.palantir.gradle.conjure:gradle-conjure')}'
+ *                 classpath '${resolve('com.palantir.sls-packaging:gradle-sls-packaging')}'
+ *             }
+ *         }
+ *
+ *         dependencies {
+ *             implementation '${resolve('com.palantir.conjure.java:conjure-lib')}'
+ *         }
+ *         """.stripIndent(true)
+ *  }
+ */
+public final class TestDepVersions {
```

Comment:
So I was thinking about this. Ideally, we'd set these versions on a system property or something on the `Test` task, so Gradle knows about these inputs to the test task. In a world where we didn't force test tasks to be run all the time, you'd need this to ensure correctness so the test task gets rerun. Fwiw I've been wondering if we can roll out the behaviour (not rerunning test tasks if build cached) somehow eventually, so we might as we do the correct thing to enable that from the get-go.

This would look like making a Gradle plugin to work out the test versions then a library that reads this from the system property. Another benefit of this approach would be we could just resolve a configuration (that also gets locked by GCV) to work out the versions we need - this would avoid needing to manually parse the versions.props/lock. Also, with this approach if you added a dep to the versions.props just for testing, GCV would say it's unused and ask you remove it. You'd have to add it to some configuration to stop the check failing (or worse disable the checkUnusedConstraints check). So we might as well add it to a configuration we use.

Reply:
My thought was that you add the dependencies used in the test as a testRuntimeOnly dependency to the test project since that is literally what they are.  That is what I've done in other projects where I have adopted this pattern already.  That avoids the need for new configurations or other plugin work.